crates/phactory/src/prpc_service.rs

```diff
@@ -463,6 +472,46 @@ impl<Platform: pal::Platform> Phactory<Platform> {
         Ok(fit_size(messages, output_buf_len))
     }
 
+    fn send_coordinate_info(&mut self, request: pb::SendCoordinateInfoRequest) -> RpcResult<()> {
+        // local ecdh_key
+        let ecdh_key = self.runtime_state()?.ecdh_key.clone();
+        // contract public ecdh_key
+        // TODO: currently assume contract key equals to local ecdh key
+        let public_contract_ecdh_key = ecdh_key.clone().public();
+        // fake key map
+        let key_map = |topic: &[u8]| {
+            Some(public_contract_ecdh_key)
+        };
+        let id_pair = self.runtime_state.as_ref()
+            .unwrap()
+            .identity_key
+            .clone();
+
+        // let sender = MessageOrigin::AccountId(id_pair.public().0.into());
+        let sender = MessageOrigin::Worker(id_pair.public());
+        let mq = self.runtime_state.as_ref()
+            .unwrap()
+            .send_mq
+            .channel(sender, id_pair);
+        let secret_mq = SecretMessageChannel::new(&ecdh_key,
+                                                  &mq,
```

Comment:
Looks like this will break the deterministic of the mq egress?

Reply:
I previously implemented a version that uses pallets to redirect contract commands, but later I with @shelvenzhou decided to refactor the code to initialize the contract command call directly in pRuntime, mainly to take advantages of SecretMessageChannel.